### PR TITLE
ARROW-11568: [C++][Compute] Rewrite mode kernel

### PR DESCRIPTION
**Pull request by cyb70289**
Arrow mode kernel performance is bad compared with scipy.stats.mode
(based on numpy.unique). Arrow mode kernel stores value:count pair in
a map, while numpy.unique sorts the input array then count the adjacent
same values. Per my test, the map approach only wins when there are
many duplicated values (length / value_range > 100), looks not very
useful in practice.

This patch rewrites mode kernel to use the sort and count approach for
floating points and integers with wide value range. 2x performance
improvement is observed.

### CODE REVIEW

**Comment by cyb70289** [comment]
Benchmark on skylake, clang-9.

```
-----------------------------------------------------------------------------------------------
Non-regressions: (40)
-----------------------------------------------------------------------------------------------
                                 benchmark         baseline        contender  change % counters
       ModeKernelWide<Int32Type>/1048576/0   20.644 MiB/sec   66.187 MiB/sec   220.613       {}
     ModeKernelWide<Int32Type>/1048576/100   20.768 MiB/sec   65.772 MiB/sec   216.699       {}
   ModeKernelWide<Int32Type>/1048576/10000   20.626 MiB/sec   65.291 MiB/sec   216.546       {}
      ModeKernelWide<Int32Type>/1048576/10   22.390 MiB/sec   69.904 MiB/sec   212.209       {}
   ModeKernelWide<FloatType>/1048576/10000   18.174 MiB/sec   54.517 MiB/sec   199.975       {}
       ModeKernelWide<FloatType>/1048576/0   18.423 MiB/sec   54.869 MiB/sec   197.826       {}
      ModeKernelWide<FloatType>/1048576/10   20.009 MiB/sec   59.346 MiB/sec   196.601       {}
     ModeKernelWide<FloatType>/1048576/100   18.486 MiB/sec   54.568 MiB/sec   195.186       {}
     ModeKernelWide<Int64Type>/1048576/100   46.593 MiB/sec  136.786 MiB/sec   193.578       {}
       ModeKernelWide<Int64Type>/1048576/0   46.434 MiB/sec  134.296 MiB/sec   189.221       {}
   ModeKernelWide<Int64Type>/1048576/10000   46.435 MiB/sec  134.149 MiB/sec   188.894       {}
     ModeKernelNarrow<Int64Type>/1048576/0  893.469 MiB/sec    2.515 GiB/sec   188.197       {}
 ModeKernelNarrow<Int64Type>/1048576/10000  886.403 MiB/sec    2.484 GiB/sec   186.933       {}
      ModeKernelWide<Int64Type>/1048576/10   50.182 MiB/sec  143.034 MiB/sec   185.029       {}
  ModeKernelWide<DoubleType>/1048576/10000   40.525 MiB/sec  115.477 MiB/sec   184.953       {}
    ModeKernelWide<DoubleType>/1048576/100   41.114 MiB/sec  116.862 MiB/sec   184.241       {}
     ModeKernelWide<DoubleType>/1048576/10   44.555 MiB/sec  126.095 MiB/sec   183.008       {}
      ModeKernelWide<DoubleType>/1048576/0   41.083 MiB/sec  113.499 MiB/sec   176.263       {}
       ModeKernelWide<FloatType>/1048576/2   39.704 MiB/sec  106.142 MiB/sec   167.335       {}
   ModeKernelNarrow<Int64Type>/1048576/100  853.217 MiB/sec    2.223 GiB/sec   166.836       {}
       ModeKernelWide<Int32Type>/1048576/2   43.329 MiB/sec  115.209 MiB/sec   165.893       {}
      ModeKernelWide<DoubleType>/1048576/2   88.485 MiB/sec  223.284 MiB/sec   152.340       {}
       ModeKernelWide<Int64Type>/1048576/2   97.572 MiB/sec  236.831 MiB/sec   142.723       {}
     ModeKernelNarrow<Int32Type>/1048576/0  715.706 MiB/sec    1.522 GiB/sec   117.703       {}
 ModeKernelNarrow<Int32Type>/1048576/10000  714.237 MiB/sec    1.497 GiB/sec   114.577       {}
    ModeKernelNarrow<Int64Type>/1048576/10  701.986 MiB/sec    1.399 GiB/sec   104.094       {}
   ModeKernelNarrow<Int32Type>/1048576/100  666.834 MiB/sec    1.291 GiB/sec    98.234       {}
     ModeKernelNarrow<Int64Type>/1048576/2  611.412 MiB/sec    1.057 GiB/sec    77.008       {}
    ModeKernelNarrow<Int32Type>/1048576/10  499.852 MiB/sec  780.344 MiB/sec    56.115       {}
     ModeKernelNarrow<Int32Type>/1048576/2  402.986 MiB/sec  562.758 MiB/sec    39.647       {}
      ModeKernelNarrow<Int8Type>/1048576/1  499.705 GiB/sec  639.311 GiB/sec    27.938       {}
      ModeKernelWide<DoubleType>/1048576/1  599.969 GiB/sec  738.218 GiB/sec    23.043       {}
     ModeKernelNarrow<Int64Type>/1048576/1  610.213 GiB/sec  743.338 GiB/sec    21.816       {}
       ModeKernelWide<Int64Type>/1048576/1  608.551 GiB/sec  735.741 GiB/sec    20.901       {}
       ModeKernelWide<Int32Type>/1048576/1  611.261 GiB/sec  727.163 GiB/sec    18.961       {}
     ModeKernelNarrow<Int32Type>/1048576/1  601.834 GiB/sec  714.017 GiB/sec    18.640       {}
       ModeKernelWide<FloatType>/1048576/1  612.628 GiB/sec  716.591 GiB/sec    16.970       {}
   ModeKernelNarrow<BooleanType>/1048576/1  628.851 GiB/sec  691.936 GiB/sec    10.032       {}
   ModeKernelNarrow<BooleanType>/1048576/0   21.709 GiB/sec   22.397 GiB/sec     3.166       {}
      ModeKernelNarrow<Int8Type>/1048576/2  293.420 MiB/sec  286.770 MiB/sec    -2.266       {}

-------------------------------------------------------------------------------------------------
Regressions: (8)
-------------------------------------------------------------------------------------------------
                                   benchmark         baseline        contender  change % counters
 ModeKernelNarrow<BooleanType>/1048576/10000    1.959 GiB/sec    1.856 GiB/sec    -5.274       {}
     ModeKernelNarrow<BooleanType>/1048576/2    1.957 GiB/sec    1.854 GiB/sec    -5.293       {}
   ModeKernelNarrow<BooleanType>/1048576/100    1.972 GiB/sec    1.866 GiB/sec    -5.374       {}
    ModeKernelNarrow<BooleanType>/1048576/10    1.974 GiB/sec    1.855 GiB/sec    -6.027       {}
       ModeKernelNarrow<Int8Type>/1048576/10  364.472 MiB/sec  342.415 MiB/sec    -6.052       {}
        ModeKernelNarrow<Int8Type>/1048576/0  484.491 MiB/sec  440.347 MiB/sec    -9.111       {}
      ModeKernelNarrow<Int8Type>/1048576/100  458.333 MiB/sec  416.045 MiB/sec    -9.226       {}
    ModeKernelNarrow<Int8Type>/1048576/10000  486.679 MiB/sec  439.186 MiB/sec    -9.759       {}
```

**Comment by pitrou** [comment]
Benchmarks on Zen 2, clang 10.
```
----------------------------------------------------------------------------------------------
Non-regressions: (44)
----------------------------------------------------------------------------------------------
                                  benchmark        baseline       contender  change % counters
  ModeKernelNarrow<Int64Type>/1048576/10000   1.344 GiB/sec   3.077 GiB/sec   129.033       {}
        ModeKernelWide<FloatType>/1048576/0  26.692 MiB/sec  60.554 MiB/sec   126.860       {}
      ModeKernelWide<FloatType>/1048576/100  26.720 MiB/sec  60.031 MiB/sec   124.667       {}
    ModeKernelWide<FloatType>/1048576/10000  26.707 MiB/sec  59.770 MiB/sec   123.799       {}
        ModeKernelWide<Int32Type>/1048576/0  35.031 MiB/sec  77.918 MiB/sec   122.426       {}
      ModeKernelNarrow<Int64Type>/1048576/0   1.399 GiB/sec   3.091 GiB/sec   120.934       {}
    ModeKernelWide<Int32Type>/1048576/10000  34.942 MiB/sec  77.128 MiB/sec   120.729       {}
     ModeKernelWide<DoubleType>/1048576/100  57.929 MiB/sec 127.702 MiB/sec   120.447       {}
    ModeKernelNarrow<Int64Type>/1048576/100   1.328 GiB/sec   2.874 GiB/sec   116.482       {}
   ModeKernelWide<DoubleType>/1048576/10000  58.316 MiB/sec 125.540 MiB/sec   115.276       {}
       ModeKernelWide<DoubleType>/1048576/0  58.070 MiB/sec 122.623 MiB/sec   111.164       {}
        ModeKernelWide<FloatType>/1048576/2  55.623 MiB/sec 117.360 MiB/sec   110.994       {}
       ModeKernelWide<DoubleType>/1048576/2 116.593 MiB/sec 243.300 MiB/sec   108.675       {}
      ModeKernelWide<Int64Type>/1048576/100  79.077 MiB/sec 162.463 MiB/sec   105.449       {}
    ModeKernelWide<Int64Type>/1048576/10000  78.890 MiB/sec 158.923 MiB/sec   101.450       {}
     ModeKernelNarrow<Int64Type>/1048576/10 994.277 MiB/sec   1.895 GiB/sec    95.196       {}
        ModeKernelWide<Int32Type>/1048576/2  70.771 MiB/sec 136.449 MiB/sec    92.804       {}
        ModeKernelWide<Int64Type>/1048576/0  79.140 MiB/sec 150.926 MiB/sec    90.709       {}
      ModeKernelWide<DoubleType>/1048576/10  72.236 MiB/sec 136.781 MiB/sec    89.354       {}
        ModeKernelWide<Int64Type>/1048576/2 149.834 MiB/sec 282.411 MiB/sec    88.482       {}
       ModeKernelWide<FloatType>/1048576/10  34.595 MiB/sec  65.197 MiB/sec    88.457       {}
       ModeKernelWide<Int32Type>/1048576/10  44.789 MiB/sec  82.908 MiB/sec    85.107       {}
      ModeKernelWide<Int32Type>/1048576/100  41.497 MiB/sec  76.746 MiB/sec    84.946       {}
      ModeKernelNarrow<Int64Type>/1048576/2 799.410 MiB/sec   1.407 GiB/sec    80.226       {}
       ModeKernelWide<Int64Type>/1048576/10  94.448 MiB/sec 168.857 MiB/sec    78.782       {}
      ModeKernelNarrow<Int32Type>/1048576/0   1.099 GiB/sec   1.709 GiB/sec    55.441       {}
     ModeKernelNarrow<Int32Type>/1048576/10 700.749 MiB/sec   1.059 GiB/sec    54.802       {}
      ModeKernelNarrow<Int32Type>/1048576/2 500.474 MiB/sec 727.773 MiB/sec    45.417       {}
    ModeKernelNarrow<Int32Type>/1048576/100   1.083 GiB/sec   1.529 GiB/sec    41.151       {}
  ModeKernelNarrow<Int32Type>/1048576/10000   1.258 GiB/sec   1.658 GiB/sec    31.790       {}
       ModeKernelNarrow<Int8Type>/1048576/1 504.487 GiB/sec 625.852 GiB/sec    24.057       {}
        ModeKernelWide<Int32Type>/1048576/1 584.030 GiB/sec 704.471 GiB/sec    20.622       {}
      ModeKernelNarrow<Int32Type>/1048576/1 595.824 GiB/sec 712.818 GiB/sec    19.636       {}
      ModeKernelNarrow<Int64Type>/1048576/1 597.643 GiB/sec 707.585 GiB/sec    18.396       {}
        ModeKernelWide<FloatType>/1048576/1 582.116 GiB/sec 679.811 GiB/sec    16.783       {}
       ModeKernelWide<DoubleType>/1048576/1 583.923 GiB/sec 674.269 GiB/sec    15.472       {}
        ModeKernelWide<Int64Type>/1048576/1 590.544 GiB/sec 678.023 GiB/sec    14.813       {}
    ModeKernelNarrow<BooleanType>/1048576/1 593.884 GiB/sec 648.047 GiB/sec     9.120       {}
    ModeKernelNarrow<BooleanType>/1048576/0  36.364 GiB/sec  38.774 GiB/sec     6.626       {}
       ModeKernelNarrow<Int8Type>/1048576/2 366.079 MiB/sec 378.478 MiB/sec     3.387       {}
   ModeKernelNarrow<BooleanType>/1048576/10   2.393 GiB/sec   2.412 GiB/sec     0.784       {}
ModeKernelNarrow<BooleanType>/1048576/10000   2.392 GiB/sec   2.395 GiB/sec     0.115       {}
    ModeKernelNarrow<BooleanType>/1048576/2   2.393 GiB/sec   2.391 GiB/sec    -0.071       {}
  ModeKernelNarrow<BooleanType>/1048576/100   2.395 GiB/sec   2.369 GiB/sec    -1.084       {}

-------------------------------------------------------------------------------------------
Regressions: (4)
-------------------------------------------------------------------------------------------
                               benchmark        baseline       contender  change % counters
    ModeKernelNarrow<Int8Type>/1048576/0 517.607 MiB/sec 477.913 MiB/sec    -7.669       {}
ModeKernelNarrow<Int8Type>/1048576/10000 597.905 MiB/sec 473.417 MiB/sec   -20.821       {}
   ModeKernelNarrow<Int8Type>/1048576/10 616.529 MiB/sec 486.274 MiB/sec   -21.127       {}
  ModeKernelNarrow<Int8Type>/1048576/100 652.528 MiB/sec 478.726 MiB/sec   -26.635       {}
```

File: cpp/src/arrow/compute/kernels/aggregate_mode.cc [hunk 1]
```diff
@@ -31,340 +33,359 @@ namespace internal {
...
+      for (const auto& array : datum.chunks()) {
+        const ArrayData& data = *array->data();
+        const CType* values = data.GetValues<CType>(1);
+        VisitSetBitRunsVoid(data.buffers[0], data.offset, data.length,
+                            [&](int64_t pos, int64_t len) {
+                              for (int64_t i = 0; i < len; ++i) {
+                                min = std::min(min, values[pos + i]);
+                                max = std::max(max, values[pos + i]);
+                              }
+                            });
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/aggregate_mode.cc:305]: It looks like this operation (compute the minmax of a Datum or ArrayData) is common enough to be factored out in an internal header file?

File: cpp/src/arrow/compute/kernels/aggregate_mode.cc [hunk 2]
```diff
@@ -31,340 +33,359 @@ namespace internal {
...
+template <typename _, typename InType>
+struct ModeExecutor {
+  static void Exec(KernelContext* ctx, const ExecBatch& batch, Datum* out) {
+    if (ctx->state() == nullptr) {
+      ctx->SetStatus(Status::Invalid("Mode requires ModeOptions"));
+      return;
+    }
+    const ModeOptions& options = ModeState::Get(ctx);
+    if (options.n <= 0) {
+      ctx->SetStatus(Status::Invalid("ModeOption::n must be positive"));
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/aggregate_mode.cc:358]: "strictly positive" or "> 0"

**Comment by pitrou** [comment]
Do the existing tests exercise both the narrow and wide cases?

**Comment by cyb70289** [comment]
> Do the existing tests exercise both the narrow and wide cases?

Yes. https://github.com/apache/arrow/blob/master/cpp/src/arrow/compute/kernels/aggregate_test.cc#L1075-L1082
I do find a comment should be updated `hashmap-based` -> `sorter-based`

File: cpp/src/arrow/compute/kernels/aggregate_mode.cc (hunk 2 above)
Comment (cyb70289) [reply cpp/src/arrow/compute/kernels/aggregate_mode.cc:358]: Done

File: cpp/src/arrow/compute/kernels/aggregate_mode.cc (hunk 1 above)
Comment (cyb70289) [reply cpp/src/arrow/compute/kernels/aggregate_mode.cc:305]: Will do.
Quantile and Mode kernels share many common code. Will also factor.

File: cpp/src/arrow/util/bit_run_reader.h [hunk 3]
```diff
@@ -480,8 +480,8 @@ Status VisitSetBitRuns(const uint8_t* bitmap, int64_t offset, int64_t length,
 }
 
 template <typename Visit>
-void VisitSetBitRunsVoid(const uint8_t* bitmap, int64_t offset, int64_t length,
-                         Visit&& visit) {
+inline void VisitSetBitRunsVoid(const uint8_t* bitmap, int64_t offset, int64_t length,
```
Comment (cyb70289) [inline cpp/src/arrow/util/bit_run_reader.h:483]: Add `inline` hint.
If the caller is from a cpp source, compiler is willing to inline. But if the caller is from a header file, compiler prefers non-inline, though in reality it doesn't increase binary size compared with called from source. Non-inline causes big perf drop as the visitor becomes a function call and cannot be optimized together with the loop.

File: cpp/src/arrow/compute/kernels/vector_sort.cc [hunk 4]
```diff
@@ -492,16 +493,8 @@ class ArrayCountOrCompareSorter {
...
-
-      VisitRawValuesInline(
-          values,
-          [&](c_type v) {
-            min = std::min(min, v);
-            max = std::max(max, v);
-          },
-          []() {});
+      c_type min, max;
+      std::tie(min, max) = GetMinMax<c_type>(*values.data());
```
Comment (cyb70289) [inline cpp/src/arrow/compute/kernels/vector_sort.cc:497]: A bit performance improvement for int64narrow sorting.
```
                           benchmark            baseline           contender  change %
 ArraySortIndicesInt64Narrow/32768/2     507.257 MiB/sec     632.995 MiB/sec    24.788
ArraySortIndicesInt64Narrow/32768/10     643.182 MiB/sec     724.483 MiB/sec    12.640
```

File: cpp/src/arrow/compute/kernels/aggregate_mode.cc (hunk 1 above)
Comment (cyb70289) [reply cpp/src/arrow/compute/kernels/aggregate_mode.cc]: Done

**Review by pitrou — APPROVED** [review]
+1, thank you!

**Comment by pitrou** [comment]
I checked Github Actions and Travis-CI on my fork, will merge.

**Pull request closed without merging**